pkg/upgrade/jiva.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2020-2021 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/openebs/openebsctl/pkg/client"
+	"github.com/openebs/openebsctl/pkg/util"
+	batchV1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+type jivaUpdateConfig struct {
+	fromVersion        string
+	toVersion          string
+	namespace          string
+	pvNames            []string
+	backOffLimit       int32
+	serviceAccountName string
+	logLevel           int32
+}
+
+// Jiva Data-plane Upgrade Job instantiator
+func InstantiateJivaUpgrade(openebsNs string, toVersion string, menifestFile string) {
+	k, err := client.NewK8sClient("")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error creating k8s client")
+		return
+	}
+
+	// If manifest Files is provided, apply the file to create a new upgrade-job
+	if menifestFile != "" {
+		yamlFile, err := yamlToJobSpec(menifestFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error in Job: %s", err)
+		}
+		k.CreateBatchJob(yamlFile)
+		return
+	}
+
+	volNames, fromVersion, desiredVersion, err := GetJivaVolumes(k)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	if toVersion == "" {
+		if desiredVersion != fromVersion {
+			// Mark it as toVersion
+			toVersion = desiredVersion
+		} else {
+			// TODO: Upgrade version to latest available version for Jiva volumes
+			fmt.Println("Fetching latest version from the remote")
+			// ...
```

Comment:
How Can I get latest version of Jiva?
I was looking at github repo for jiva operator to get some yaml from where I could extract it.

Reply:
Got it now

Reply:
Actually the version to update depends upon the version of `openebs/uprade` since we are using the same to version as the upgrade image.
from yaml in the jiva upgrade volume: [:arrow_down: <- ](https://github.com/openebs/upgrade/blob/develop/docs/upgrade.md#running-the-upgrade-job-2)
```
        # the image version should be same as the --to-version mentioned above
        # in the args of the job
        image: openebs/upgrade:<same-as-to-version>
        imagePullPolicy: IfNotPresent
```